sqlalchemy-stubs/sql/functions.pyi

```diff
@@ -1,250 +1,308 @@
-from .elements import ColumnElement as ColumnElement
-from .elements import Extract as Extract
-from .elements import FunctionFilter as FunctionFilter
-from .elements import Grouping as Grouping
-from .elements import literal_column as literal_column
-from .elements import NamedColumn as NamedColumn
-from .elements import Over as Over
-from .elements import WithinGroup as WithinGroup
-from .selectable import FromClause as FromClause
-from .selectable import Select as Select
-from .selectable import TableValuedAlias as TableValuedAlias
-from .visitors import InternalTraversal as InternalTraversal
-from .visitors import TraversibleType as TraversibleType
-from .. import util as util
+from typing import overload
+from typing import Type
+from typing import TypeVar
+from typing import Union
+
+from . import sqltypes
+from . import type_api
+from .base import ColumnCollection
+from .base import Executable
+from .base import Generative
+from .base import HasMemoized
+from .elements import AsBoolean
+from .elements import BinaryExpression
+from .elements import ClauseElement
+from .elements import ClauseList
+from .elements import ColumnElement
+from .elements import FunctionFilter
+from .elements import Grouping
+from .elements import NamedColumn
+from .elements import Over
+from .elements import TableValuedColumn
+from .elements import WithinGroup
+from .schema import Sequence
+from .selectable import FromClause
+from .selectable import Select
+from .selectable import TableValuedAlias
+from .visitors import TraversibleType
+
+_TE = TypeVar("_TE", bound=type_api.TypeEngine[Any])
+_FE = TypeVar("_FE", bound=FunctionElement[Any])
 
 def register_function(
-    identifier: Any, fn: Any, package: str = ...
+    identifier: str, fn: Any, package: str = ...
 ) -> None: ...
 
-class FunctionElement(Executable, ColumnElement, FromClause, Generative):
+class FunctionElement(  # type: ignore[misc]
+    Executable, ColumnElement[_TE], FromClause, Generative, Generic[_TE]
+):
     packagenames: Any = ...
     clause_expr: Any = ...
     def __init__(self, *clauses: Any, **kwargs: Any) -> None: ...
-    def scalar_table_valued(self, name: Any, type_: Optional[Any] = ...): ...
-    def table_valued(self, *expr: Any, **kw: Any): ...
-    def column_valued(self, name: Optional[Any] = ...): ...
+    @overload
+    def scalar_table_valued(
+        self, name: Any, type_: None = ...
+    ) -> ScalarFunctionColumn[sqltypes.NullType]: ...
+    @overload
+    def scalar_table_valued(
+        self, name: Any, type_: Union[_TE, Type[_TE]]
+    ) -> ScalarFunctionColumn[_TE]: ...
+    def table_valued(self, *expr: Any, **kw: Any) -> TableValuedAlias: ...  # type: ignore[override]
+    def column_valued(
+        self, name: Optional[Any] = ...
+    ) -> TableValuedColumn[Any]: ...
     @property
-    def columns(self): ...
-    def clauses(self): ...
+    def columns(self) -> ColumnCollection[Any]: ...  # type: ignore[override]
```

Comment:
I think we settled on using `ColumnCollection[ColumnElement[Any]]` for the selects, so I guess this should follow?